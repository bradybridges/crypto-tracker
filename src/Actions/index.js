export const updateCryptos = (cryptos) => ({
  type: 'UPDATE_CRYPTOS',
  cryptos,
});

export const updateError = (error) => ({
  type: 'UPDATE_ERROR',
  error,
});

export const updateTimeFrame = (timeFrame) => ({
  type: 'UPDATE_TIME_FRAME',
  timeFrame,
});

export const updateTrackedCoins = (symbols) => ({
  type: 'UPDATE_TRACKED_COINS',
  symbols,
});

export const updatePortfolio = (portfolio) => ({
  type: 'UPDATE_PORTFOLIO',
  portfolio,
});
