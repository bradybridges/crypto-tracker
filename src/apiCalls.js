import { apiKey } from './apiKey';

export const fetchTopCryptos = async () => {
  const url = 'https://api.nomics.com/v1/currencies/ticker';
  const key = `?key=${apiKey}`;
  const queries = '&ids=BTC,ETH,XRP,LTC,XMR,ETC,BCH&interval=1h,1d,30d,365d&convert=USD'
  const response = await fetch(`${url}${key}${queries}`);
  if(!response.ok) {
    throw Error('Failed to fetch coins...');
  } else {
    const coins = await response.json();
    return coins;
  }
}
