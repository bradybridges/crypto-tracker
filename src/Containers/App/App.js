import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Route } from 'react-router-dom';
import { fetchTopCryptos } from '../../apiCalls';
import { connect } from 'react-redux';
import { updateCryptos, updateError, updateTrackedCoins, updatePortfolio } from '../../Actions/';
import Header from '../../Components/Header/Header';
import CryptoContainer from '../CryptoContainer/CryptoContainer';
import Nav from '../../Components/Nav/Nav';
import SearchCoins from '../SearchCoins/SearchCoins';
import Portfolio from '../Portfolio/Portfolio';
import CoinInfo from '../CoinInfo/CoinInfo';
import './App.scss';

export class App extends Component {

  componentDidMount = () => {
    const localTrackedCoins = JSON.parse(localStorage.getItem('trackedCoins'));
    if(localTrackedCoins) {
      this.props.updateTrackedCoins(localTrackedCoins);
      this.getCryptos(localTrackedCoins);
      this.loadPortfolio();
    } else {
      this.getCryptos(this.props.trackedCoins);
    }
  }

  loadPortfolio = () => {
    const localPortfolio = JSON.parse(localStorage.getItem('portfolio'));

    if(localPortfolio) {
      this.props.updatePortfolio(localPortfolio);
    }
  }

  getCryptos = symbols => {
    fetchTopCryptos(symbols)
    .then(cryptos => {
      this.props.updateCryptos(cryptos);
    })
    .catch(error => {
      this.props.updateError(error.message);
    });
  }

  handleCryptoRefresh = () => {
    setInterval(() => {
      this.getCryptos();
    }, 60000);
  }

  render() {
    // this.handleCryptoRefresh(); //uncomment for 60 sec refresh of data
    return (
      <main>
        <Header />
        <Route exact path='/' component={CryptoContainer} />
        <Route path='/search' render={() => <SearchCoins getCryptos={this.getCryptos} />} />
        <Route path='/portfolio' component={Portfolio} />
        <Route path='/coins/:name' render={({ match }) => {
	        const { name } = match.params;
	        const coin = this.props.cryptos.find(coin => coin.name === name);
          return coin && <CoinInfo coin={coin} />;
        }} />
        <nav>
          <Nav />
        </nav>
      </main>
      
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  updateCryptos: cryptos => dispatch( updateCryptos(cryptos) ),
  updateError: error => dispatch( updateError(error) ),
  updateTrackedCoins: coins => dispatch( updateTrackedCoins(coins) ),
  updatePortfolio: portfolio => dispatch( updatePortfolio(portfolio) ),
});

export const mapStateToProps = state => ({
  cryptos: state.cryptos,
  error: state.error,
  trackedCoins: state.trackedCoins,
});

export default connect(mapStateToProps, mapDispatchToProps)(App);

App.propTypes = {
  updateCryptos: PropTypes.func.isRequired,
  updateError: PropTypes.func.isRequired,
  updateTrackedCoins: PropTypes.func.isRequired,
  updatePortfolio: PropTypes.func.isRequired,
  cryptos: PropTypes.array.isRequired,
  trackedCoins: PropTypes.array.isRequired,
};
