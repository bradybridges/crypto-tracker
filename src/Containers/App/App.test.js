import React from 'react';
import { shallow } from 'enzyme';
import { App, mapDispatchToProps, mapStateToProps } from './App';
import * as api from '../../apiCalls';
import * as actions from '../../Actions/index';

describe('App', () => {
  let wrapper;
  const mockCryptos = [
    {
      currecy: 'BTC',
      id: 'BTC',
      price: 9341.6987,
      price_date: '2019-11-02',
      symbol: 'BTC',
      circulating_supply: '18025852',
      max_supply: '21000000',
      name: 'Bitcoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg',
      market_cap: '1243423423.05',
      rank: '1',
      high: '19480.90',
      high_timestamp: '2017-12-16',
    },
    {
      currecy: 'LTC',
      id: 'LTC',
      price: 75.55,
      price_date: '2019-11-02',
      symbol: 'LTC',
      circulating_supply: '18000000',
      max_supply: '20000000',
      name: 'Litecoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/ltc.svg',
      market_cap: '1003423423.05',
      rank: '3',
      high: '420.90',
      high_timestamp: '2017-12-10',
    },
  ];

  beforeEach(() => {
    wrapper = shallow(
      <App 
        cryptos={mockCryptos}
        error=''
        trackedCoins={['BTC', 'LTC']}
        updateCryptos={jest.fn()}
        updateError={jest.fn()}
        updateTrackedCoins={jest.fn()}
        updatePortfolio={jest.fn()}
      />
    ); 
  }); 

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('getCryptos should call fetchTopCryptos and updateCryptos', () => {
    api.fetchTopCryptos = jest.fn().mockImplementation(() => {
      return Promise.resolve([{name: 'Bitcoin'}]);
    });
    wrapper.instance().getCryptos();
    expect(api.fetchTopCryptos).toHaveBeenCalled();
    // expect(wrapper.instance().props.updateCryptos).toHaveBeenCalled();
  })
});

describe('mapStateToProps', () => {
  const mockCryptos = [ 
    {
      currecy: 'BTC',
      id: 'BTC',
      price: 9341.6987,
      price_date: '2019-11-02',
      symbol: 'BTC',
      circulating_supply: '18025852',
      max_supply: '21000000',
      name: 'Bitcoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/btc.svg',
      market_cap: '1243423423.05',
      rank: '1',
      high: '19480.90',
      high_timestamp: '2017-12-16',
    },
    {
      currecy: 'LTC',
      id: 'LTC',
      price: 75.55,
      price_date: '2019-11-02',
      symbol: 'LTC',
      circulating_supply: '18000000',
      max_supply: '20000000',
      name: 'Litecoin',
      logo_url: 'https://s3.us-east-2.amazonaws.com/nomics-api/static/images/currencies/ltc.svg',
      market_cap: '1003423423.05',
      rank: '3',
      high: '420.90',
      high_timestamp: '2017-12-10',
    },
  ];

  it('should return cryptos, error, and trackedCoins', () => {
    const mockState = {
      cryptos: mockCryptos,
      error: 'An Error Occured',
      trackedCoins: ['BTC', 'LTC'],
      portfolio: [{name: 'Bitcoin', qty: 10}, {name: 'Litecoin', qty: 20}],
    }

    const expected = {
      cryptos: mockCryptos,
      error: 'An Error Occured',
      trackedCoins: ['BTC', 'LTC'],
    }

    const result = mapStateToProps(mockState);
    expect(result).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  let mockDispatch, mappedDispatch;
  
  beforeEach(() => {
    mockDispatch = jest.fn();
    mappedDispatch = mapDispatchToProps(mockDispatch);
  });

  it('updateCrytos is dispatched with correct action', () => {
    const actionToDispatch = actions.updateCryptos([{name: 'Bitcoin'}]);
    mappedDispatch.updateCryptos([{name: 'Bitcoin'}]);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('updateError is dispatched with correct action', () => {
    const actionToDispatch = actions.updateError('Error occured');
    mappedDispatch.updateError('Error occured');
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('updateTrackedCoins should be dispatched with correct action', () => {
    const actionToDispatch = actions.updateTrackedCoins(['BTC','LTC']);
    mappedDispatch.updateTrackedCoins(['BTC', 'LTC']);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('updatePortfolio should be dispatched with correct action', () => {
    const mockPortfolio = [
      {name: 'Bitcoin', qty: 10},
      {name: 'Litecoin', qty: 50},
    ];
    const actionToDispatch = actions.updatePortfolio(mockPortfolio);
    mappedDispatch.updatePortfolio(mockPortfolio);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});