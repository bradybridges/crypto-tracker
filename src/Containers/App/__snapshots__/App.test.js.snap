// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<main>
  <Header />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(CryptoContainer)",
        "type": [Function],
      }
    }
    exact={true}
    path="/"
  />
  <Route
    path="/search"
    render={[Function]}
  />
  <Route
    component={
      Object {
        "$$typeof": Symbol(react.memo),
        "WrappedComponent": [Function],
        "compare": null,
        "displayName": "Connect(Portfolio)",
        "type": [Function],
      }
    }
    path="/portfolio"
  />
  <Route
    path="/coins/:name"
    render={[Function]}
  />
  <nav>
    <Nav />
  </nav>
</main>
`;
